x/epoching/keeper/keeper.go

```diff
@@ -0,0 +1,189 @@
+
+	bz := store.Get(ActionStoreKey(epochNumber, actionID))
+	if bz == nil {
+		return nil
+	}
+
+	var action sdk.Msg
+	// reference from TestMarshalAny(t *testing.T)
+	k.cdc.UnmarshalInterface(bz, &action)
+
+	return action
+}
+
+// GetEpochActions get all actions
+func (k Keeper) GetEpochActions(ctx sdk.Context) []sdk.Msg {
+	actions := []sdk.Msg{}
+	iterator := sdk.KVStorePrefixIterator(ctx.KVStore(k.storeKey), []byte(EpochActionQueuePrefix))
+
+	for ; iterator.Valid(); iterator.Next() {
+		var action sdk.Msg
+		bz := iterator.Value()
+		// reference from TestMarshalAny(t *testing.T)
+		k.cdc.UnmarshalInterface(bz, &action)
+		actions = append(actions, action)
+	}
+
+	return actions
+}
+
+// GetEpochActionsIterator returns iterator for EpochActions
+func (k Keeper) GetEpochActionsIterator(ctx sdk.Context) db.Iterator {
+	return sdk.KVStorePrefixIterator(ctx.KVStore(k.storeKey), []byte(EpochActionQueuePrefix))
+}
+
+// DequeueEpochActions dequeue all the actions store on epoch
+func (k Keeper) DequeueEpochActions(ctx sdk.Context) {
+	store := ctx.KVStore(k.storeKey)
+	iterator := sdk.KVStorePrefixIterator(store, []byte(EpochActionQueuePrefix))
+
+	for ; iterator.Valid(); iterator.Next() {
+		key := iterator.Key()
+		store.Delete(key)
+	}
+}
+
+// DeleteByKey delete item by key
+func (k Keeper) DeleteByKey(ctx sdk.Context, key []byte) {
+	store := ctx.KVStore(k.storeKey)
+	store.Delete(key)
+}
+
+// GetEpochActionByIterator get action by iterator
+func (k Keeper) GetEpochActionByIterator(iterator db.Iterator) sdk.Msg {
+	bz := iterator.Value()
+
+	var action sdk.Msg
+	// reference from TestMarshalAny(t *testing.T)
+	k.cdc.UnmarshalInterface(bz, &action)
+
+	return action
+}
+
+// SetEpochNumber set epoch number
+func (k Keeper) SetEpochNumber(ctx sdk.Context, epochNumber int64) {
+	store := ctx.KVStore(k.storeKey)
+	store.Set(EpochNumberID, sdk.Uint64ToBigEndian(uint64(epochNumber)))
+}
+
+// GetEpochNumber fetches epoch number
+func (k Keeper) GetEpochNumber(ctx sdk.Context) int64 {
+	store := ctx.KVStore(k.storeKey)
+
+	bz := store.Get(EpochNumberID)
+	if bz == nil {
+		// return default EpochNumber 0
+		return 0
+	}
+
+	return int64(sdk.BigEndianToUint64(bz))
+}
+
+// IncreaseEpochNumber increases epoch number
+func (k Keeper) IncreaseEpochNumber(ctx sdk.Context) {
+	epochNumber := k.GetEpochNumber(ctx)
+	k.SetEpochNumber(ctx, epochNumber+1)
+}
+
+// GetNextEpochHeight returns next epoch block height
+func (k Keeper) GetNextEpochHeight(ctx sdk.Context, epochInterval int64) int64 {
+	currentHeight := ctx.BlockHeight()
+	return currentHeight + (epochInterval - currentHeight%epochInterval)
+}
+
+// GetNextEpochTime returns estimated next epoch time
+func (k Keeper) GetNextEpochTime(ctx sdk.Context, epochInterval int64) time.Time {
+	currentTime := ctx.BlockTime()
+	currentHeight := ctx.BlockHeight()
+	timeoutCommit := 5 * time.Second // TODO how to get timeout commit tendermint config?
+	// cp := baseapp.GetConsensusParams(ctx)
```

Comment:
This is going to be problematic to assume. Validators can change this locally to skew block production (we actually did this recently on the hub to get a proposal executed in time).

Can we instead pass a parameter, `commitTimeout`, to the `Keeper`? The value would be fetched from config in `app.go`.